### ISSUE TITLE
Accessibility conformance Testing (ACT) Rules Format 1.1

### ISSUE THREAD

**Issue by daniel-montalvo** [w3ctag/design-reviews#977]
こんにちは TAG-さん!

I'm requesting a TAG review of Accessibility Conformance Testing (ACT) Rules Format 1.1.

The purpose of the Accessibility Conformance Testing (ACT) effort is to establish and document rules for testing the conformance of web content to accessibility standards, such as Web Content Accessibility Guidelines (WCAG). These test rules address automated, semi-automated, and manual testing. ACT makes accessibility testing more transparent, and thus reduces confusion caused by different interpretations of accessibility guidelines.

  - Explainer¹ (minimally containing user needs and example code): https://www.w3.org/WAI/standards-guidelines/act/
  - Specification URL: https://www.w3.org/TR/2024/WD-act-rules-format-1.1-20240618/
  - Tests: NA
  - User research: none
  - Security and Privacy self-review²: none
  - GitHub repo: https://www.github.com/w3c/wcag-act/
  - Primary contacts (and their relationship to the specification):
      - [name] ([github username]), [organization/s] (repeat as necessary, we recommend including group chairs and editors in this list)
      - Daniel Montalvo, daniel-montalvo, W3C, Staff Contact
      - Wilco Fiers, wilcofiers, Deque, Task Force co-Facilitator
      - Katherine Eng, kengdoj, Access Board, Task Force co-Facilitator
  - Organization(s)/project(s) driving the specification: Deque, SiteImprove, Access Board, IBMproject name]
  - Key pieces of existing multi-stakeholder (e.g. developers, implementers, civil society) support, review or discussion of this specification
  - Key pieces of multi-implementer support:
    - Chromium comments: [url]
    - Mozilla comments: https://github.com/mozilla/standards-positions/issues/NNN
    - WebKit comments: https://github.com/WebKit/standards-positions/issues/NNN
    - etc...
  - External status/issue trackers for this specification (publicly visible, e.g. Chrome Status):

Further details:

  - [x] I have reviewed the TAG's [Web Platform Design Principles](https://www.w3.org/TR/design-principles/)
  - Relevant time constraints or deadlines: Ideally you can provide comments in two months from now.
  - The group where the work on this specification is currently being done: Accessibility Conformance Testing Task Force
  - The group where standardization of this work is intended to be done (if current group is a community group or other incubation venue): Accessibility Guidelines Working Group
  - Major unresolved issues with or opposition to this specification:
  - This work is being funded by:

You should also know that...

[please tell us anything you think is relevant to this review]

**Comment by matatk** [comment]
Hi @daniel-montalvo, we are following up on the ACT Rules review this week. We were wondering if any of the avenues suggested above align with the group's plans for this work? If there is any advice on how to proceed along those directions that we can offer, please let us know.

**Comment by daniel-montalvo** [comment]
Sorry @matatk this slipped through.

I'll respond below.

> * The status quo: this document is specifying a format for writing ACT rules—in which case, the ACT rules documents themselves would count as implementations. We encourage you to ensure that the specification is precise enough to allow tools to be written that could verify, or 'lint', ACT rules against this spec—these linting tools would then form a test suite that ACT Rules authors could use to verify the rules documents they're writing.

In our view, commonly used accessibility checkers are already doing this job, although it'd be desirable that more vendors will join.

Currently, the process is as follows:

- The rules are considered "implementations" of this format specification.
- Everybody can write rules that conform with the format specification.
- There is a [set of rules](https://www.w3.org/WAI/standards-guidelines/act/rules/) written by the ACT Task Force. Some of them have already been approved by AGWG, others are proposed by the Task Force and will eventually be approved by the Working Group.

Although implementation guidance is always welcome, wouldn't be out of scope for the rules format document itself?

> * We also encourage you to publish a REC track document that aggregates the ACT rules as written by the CG—in which case the requirement for _them_ to mature on the REC track would be that accessibility testing tools embodying (or supporting) the rules would be required.

I am not clear on this point. Doesn't the  [set of rules](https://www.w3.org/WAI/standards-guidelines/act/rules/) mentioned above cover this already? In addition to this, there is also [Implementation pages](https://www.w3.org/WAI/standards-guidelines/act/implementations/) that cover how different automated, semi-automated, and manual tools have indeed implemented the rule set.

It's true these are not on Rec track because back in the day it was decided that they'll be more impactful if they're published as resources on the WAI website.

> * If you have a goal that manual testing tools should be able to load ACT rules, for humans to use in test procedures (which we encourage) then implementations would be accessibility testing tools that support the loading of ACT rules.
> * Some of the ACT rules could be checked entirely mechanically. It's possible that ways could be developed to achieve this that could be used (by a machine) directly, as part of projects such as [Playwright](https://playwright.dev/), [Cypress](https://www.cypress.io/), and/or [Web Platform Tests](https://web-platform-tests.org/) (which has a somewhat different focus). This would allow the rules (or rather the mechanical checks underpinning them) to be more widely adopted, so we encourage you to investigate this approach in future.

- This is certainly something that could be developed in the future. We have not explored yet how the rules could be used in the context of Playwright or Cipress.
- The group is currently exploring how test cases within rules can contribute to the current accessibility Web Platform Tests. For the moment it's mostly automated, manual, and semi-automated accessibility checkers which are implementing the test cases in the rules.
- There is also work on manual rules (rules that automated checkers cannot currently check by themselves) but these are still not approved by the Task Force, hence not published as Proposed to the AGWG Working Group.

The Task Force Would appreciate any comments if you think these additional pages and resource still don't cover the points you raise.

**Comment by matatk** [comment]
Thanks for your detailed and helpful reply @daniel-montalvo - this gives us a clear picture of the goals, and this gives rise to some things we'd like you to consider.

As the rules are implementations of the spec (the rules format) it's important that the format be made sufficiently precise such that the rules themselves can be automatically checked against the format. We imagine a linting program would be very helpful to those in the community writing the rules. Here are a few examples (which are not exhaustive) as to how that could be achieved:

* Prescribing a single format in which the rules be written would simplify the process of checking them, and mean probably only one linting program would need to be written.

* Here are some examples of things that would need to be tightened up in the spec in order to make automated linting possible:

  1. The exact names for section headings.

  2. Possibly also the ordering for section headings (though this would be more to keep output consistent for readers).

  3. Specifying the exact strings to use as enum values. E.g. for [rule types](https://w3c.github.io/wcag-act/act-rules-format.html#rule-type) what would be the exact text string that indicates the rule is either atomic or composite?

  4. Specifying the exact structure - and permitted values - of the [accessibility requirements mapping](https://w3c.github.io/wcag-act/act-rules-format.html#accessibility-requirements-mapping) - i.e. how are the 5 fields identified, and what are the allowed values (exact text strings)?

  5. Could expectations ([example expectations section](https://github.com/w3c/wcag-act-rules/blob/6e8ff64e66067a7e2304b9e051055151d9be6493/content/rules/73f2c2/index.md?plain=1#L40-L48)) be more tightly specified, such that they could be (in future) loaded into an automated test framework? It seems that this would be possible, and could have significant benefits later on.

Your explanation as to why you are not pursuing REC track for the rules themselves makes sense. You could always pursue it later if it does make sense for the group.

Regarding the possibility of integrating the rules into general testing tools later, including WPT, we're pleased that you're exploring this. Let us know if we can help.

We hope this helps clarify things - we are keen to hear your thoughts, and provide any further clarification or advice that we can on this, so we will keep this thread open.

**Comment by daniel-montalvo** [comment]
Thanks  for your suggestions about this, @matatk.

This is indeed a great idea. We already have a good number of tests in the [ACT-rules.github.io](https://github.com/act-rules/act-rules.github.io/) repo.
Going with strings only as a way to make these normative may raise some i18n concerns. It seems the only way to go about this would be coming up with normative prose semantically, for example through an RDF schema, which is quite an undertaking given current group resources.

We are assuming this is not a blocker for the current 1.1 version. This overall sounds like a good idea to pick up in a future ACT Rules Format version, and we’d be willing to explore potential approaches for this when the time comes for a 2.0 version.

**Comment by WilcoFiers** [comment]
@matatk Have you had a chance to look at Daniel's response? We would like to have confirmation that TAG is on board with this approach before we process to CR.

**Comment by jyasskin** [comment]
We talked about this in our breakout this week: We are supportive of this work, and we are happy for the 1.1 version to go to CR.

We're pleased that you're keen to explore the specificity issues that we raised in a future revision (ideally the next after 1.1). As the spec you're writing is about the _format_ of the rules, then rule documents themselves don't constitute tests of the spec: a linter that can check any given rule documents _against the format_ would constitute a test for the _format's spec_.

We agree that it's important to ensure there are no internationalization concerns. There are ways this could be achieved whilst still fulfilling the need to make the format's spec more precise, and we'd be happy to advise on that later.

We're closing this review with the `satisfied with concerns` label in order to confirm that we support the work you're doing, and we hope to work with you in future to address the specificity issues we raised.

Thank you again for your review request, and for the important work your group is doing.
